nni/compression/pytorch/speedup/infer_mask.py

```diff
@@ -0,0 +1,331 @@
+            # ori_out may be the same tensor with dummy_input,
+            # so we use clone and detach to create a new tensor with
+            # the same values.
+            out = self.module(*self.dummy_input)
+            if isinstance(out, torch.Tensor):
+                out_mask, constant = self.isconstants(out.clone().detach())
+            elif isinstance(out, tuple) or isinstance(out, list):
+                out_mask = []
+                constant = []
+                for tout in out:
+                    _mask, _constant = self.isconstants(tout.clone().detach())
+                    out_mask.append(_mask)
+                    constant.append(_constant)
+            else:
+                _logger.warning(
+                    'Only support the OP whose output is tensor/tuple of tensor/list of tensor')
+
+        # We also need random the parameters of the module, because if the weight of the model has
+        # a unmasked 0, then our out sparsity inference may be wrong
+        # However, after radomizing the weight/parameters, the constant in the output tensors may
+        # be different from the constants that calculated from its original stata_dict. However,
+        # so to get the right constant to eliminate the bias between model before and after sparsity
+        # inference, we need to reload its state_dict and recalculate the constant
+
+        # TODO can be optimized here, move the randomization at the begining
+        if len(self.weights) > 0 and self.state_dict is not None:
+
+            self.module.load_state_dict(self.state_dict)
+            # apply weight mask
+            self.__apply_weight_mask()
+            out = self.module(*self.dummy_input).clone().detach()
+            if isinstance(out, torch.Tensor):
+                constant = torch.zeros_like(out)
+                constant_pos = out_mask == 0
+                constant[constant_pos] = out[constant_pos]
+            elif isinstance(out, (list, tuple)):
+                constant = []
+                for i, tout in enumerate(out):
+                    _tmp = torch.zeros_like(tout)
+                    sparsity_pos = out_mask[i] == 0
+                    _tmp[sparsity_pos] = tout[sparsity_pos]
+                    constant.append(_tmp)
+
+        return out_mask, constant
+        # return out_mask
+
+    def update_indirect_sparsity(self):
+        """
+        Find those hidden sparsity through gradient.
+        """
+        # Each node only update the output mask when we backwards
+        # update the output mask
+        if isinstance(self.output, torch.Tensor) and self.output.grad is not None:
+            # if output have gradient which means this node has successor
+            # nodes and the successor nodes have already update their indirect
+            # sparsity
+            # we can mask the values whose gradient is always zeros
+            gradient_sum = torch.sum(torch.abs(self.output.grad.data), dim=0)
```

Comment:
you tried to deal with different `dim` values, but `dim` is fixed to 0 here. I suggest to only support dim=0 for now. BTW, are there many cases that `dim` is not 0?